core/src/sumeragi/main_loop.rs

```diff
@@ -367,22 +363,30 @@ impl Sumeragi {
         self.connect_peers(&self.current_topology);
     }
 
-    fn update_state(&mut self, committed_block: &VersionedCommittedBlock) {
+    fn update_state(
+        &mut self,
+        committed_block: &Arc<VersionedCommittedBlock>,
+        kura_store_block: fn(&Kura, Arc<VersionedCommittedBlock>),
+    ) {
         self.wsv
-            .apply(committed_block)
+            .apply(committed_block.as_ref())
             .expect("Failed to apply block on WSV. Bailing.");
 
         self.send_events(self.wsv.events_buffer.replace(Vec::new()));
 
         // Parameters are updated before updating public copy of sumeragi
         self.update_params();
 
+        // Block must be stored by kura after updating WSV because apply block might fail and we will store corrupted block
+        // but before updating public copy of wsv and emitting BlockCommitted event to avoid situation when block is not yet stored in kura
+        kura_store_block(&self.kura, Arc::clone(committed_block));
```

Comment:
```suggestion
        // https://github.com/hyperledger/iroha/issues/3396:
        // Iroha should emit `BlockCommitted` event only after Kura stored the block.
        // However, WSV update should happen first, in order to avoid storaging a corrupted block
        kura_store_block(&self.kura, Arc::clone(committed_block));
```